src/ZkFold/Base/Protocol/ARK/Plonk.hs

```diff
@@ -3,227 +3,157 @@
-    type Proof (Plonk d n t)       = PlonkProof
-
-    setupProve :: Plonk d n t -> SetupProve (Plonk d n t)
+instance forall n l c1 c2 t plonk f g1.
+        ( Plonk n l c1 c2 t ~ plonk
+        , ScalarField c1 ~ f
+        , Point c1 ~ g1
+        , KnownNat n
+        , KnownNat l
+        , KnownNat (PlonkPermutationSize n)
+        , KnownNat (PlonkPolyExtendedLength n)
+        , Eq (ScalarField c1)
+        , Scale (ScalarField c1) (ScalarField c1)
+        , BinaryExpansion (ScalarField c1)
+        , Bits (ScalarField c1) ~ [ScalarField c1]
+        , FiniteField (ScalarField c1)
+        , AdditiveGroup (BaseField c1)
+        , Pairing c1 c2
+        , ToTranscript t (ScalarField c1)
+        , ToTranscript t (Point c1)
+        , FromTranscript t (ScalarField c1)
+        ) => NonInteractiveProof (Plonk n l c1 c2 t) where
+    type Transcript (Plonk n l c1 c2 t)  = t
+    type SetupProve (Plonk n l c1 c2 t)  = (PlonkSetupParamsProve c1 c2, PlonkPermutation n c1, PlonkCircuitPolynomials n c1 , PlonkWitnessMap n c1)
+    type SetupVerify (Plonk n l c1 c2 t) = (PlonkSetupParamsVerify c1 c2, PlonkCircuitCommitments c1)
+    type Witness (Plonk n l c1 c2 t)     = (PlonkWitnessInput c1, PlonkProverSecret c1)
+    type Input (Plonk n l c1 c2 t)       = PlonkInput c1
+    type Proof (Plonk n l c1 c2 t)       = PlonkProof c1
+
+    setupProve :: plonk -> SetupProve plonk
     setupProve plonk@(Plonk omega' k1' k2' iPub ac x) =
         (PlonkSetupParamsProve {..}, PlonkPermutation {..}, PlonkCircuitPolynomials {..}, PlonkWitnessMap $ wmap pr)
         where
-            d     = value @d + 6
+            d     = value @n + 6
             xs    = fromList $ map (x^) [0..d-!1]
             gs'   = fmap (`mul` gen) xs
             h0'   = gen
             h1'   = x `mul` gen
             iPub' = fromList . fromVector $ iPub
 
-            pr    = fromJust $ toPlonkRelation @d @n @F iPub ac
+            pr    = fromJust $ toPlonkRelation @n @l @f iPub ac
 
             perm@PlonkPermutation {..}   = plonkPermutation plonk pr
             PlonkCircuitPolynomials {..} = plonkCircuitPolynomials plonk perm pr
 
-    setupVerify :: Plonk d n t -> SetupVerify (Plonk d n t)
+    setupVerify :: plonk -> SetupVerify plonk
     setupVerify plonk@(Plonk omega k1 k2 iPub ac x) = (PlonkSetupParamsVerify {..}, PlonkCircuitCommitments {..})
         where
-            d = value @d + 6
+            d = value @n + 6
             xs = fromList $ map (x^) [0..d-!1]
             gs = fmap (`mul` gen) xs
-            g0 = V.head gs
-            h0 = gen
-            h1 = x `mul` gen
-            pow = floor @Double . logBase 2.0 . fromIntegral $ value @d
-
-            pr   = fromJust $ toPlonkRelation @d @n @F iPub ac
+            omega'' = omega
+            k1''    = k1
+            k2''    = k2
+            g0''    = gen
+            h0''    = gen
+            h1''    = x `mul` gen
+            pow''   = floor @Double . logBase 2.0 . fromIntegral $ value @n
```

Comment:
i think `floor` is a bug

Reply:
Well, `n` should be a power of two because we wouldn't be able to find a group H otherwise (at least for the BLS curve).

Reply:
I think we expect only an integer number of constraints, but for uniformity it is worth using only one method like `ceiling`